src/protocol/boolean/generate_random_bits.rs

```diff
@@ -1,153 +1,137 @@
+use std::marker::PhantomData;
+
 use crate::{
     error::Error,
-    ff::{Field, Gf40Bit, PrimeField},
+    ff::PrimeField,
     protocol::{
         basics::SecureMul,
-        context::{
-            Context, UpgradableContext, UpgradedContext, UpgradedMaliciousContext,
-            UpgradedSemiHonestContext,
+        context::{prss::InstrumentedIndexedSharedRandomness, Context, UpgradedContext},
+        modulus_conversion::{
+            convert_bits, convert_some_bits, BitConversionTriple, ToBitConversionTriples,
         },
-        modulus_conversion::{convert_bit, convert_bit_local, BitConversionTriple},
         prss::SharedRandomness,
-        step::BitOpStep,
         RecordId,
     },
     secret_sharing::{
-        replicated::{
-            malicious::{AdditiveShare as MaliciousReplicated, ExtendableField},
-            semi_honest::AdditiveShare as Replicated,
-            ReplicatedSecretSharing,
-        },
-        Linear as LinearSecretSharing, SecretSharing, SharedValue,
+        replicated::semi_honest::AdditiveShare as Replicated, BitDecomposed,
+        Linear as LinearSecretSharing,
     },
-    seq_join::SeqJoin,
 };
-use async_trait::async_trait;
-
-#[async_trait]
-pub trait RandomBits<V: SharedValue> {
-    type Share: SecretSharing<V>;
-
-    async fn generate_random_bits(self, record_id: RecordId) -> Result<Vec<Self::Share>, Error>;
+use futures::stream::{iter as stream_iter, Stream, StreamExt};
+
+#[derive(Debug)]
+struct RawRandomBits {
+    // TODO: use a const generic instead of a field, when generic_const_expr hits stable.
```

Comment:
we need a Rust compiler [feature](https://github.com/rust-lang/rust/issues/76560) that is currently available behind a feature gate an in nightly (unstable) compiler. My guess is that it is currently not possible to do the following

```rust
struct RawRandomBits<const N: u32> {
    left: u64,
    right u64
}

// and somewhere on the client side
let bits = RawRandomBits<u128::BITS - F::PRIME> { ... }
```

Reply:
@akoshelev is right.  This would be wonderful, but we can't do it just yet because the compiler isn't smart enough (yet).  But if this ever becomes const throughout, we not only save on struct size, but everything else goes faster.  And the code becomes simpler.  It would be a pretty big win.